lib/Dialect/Ibis/Transforms/IbisAddOperatorLibrary.cpp

```diff
@@ -0,0 +1,77 @@
+//===- IbisAddOperatorLibraryPass.cpp -------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+
+#include "circt/Dialect/Comb/CombOps.h"
+#include "circt/Dialect/Ibis/IbisDialect.h"
+#include "circt/Dialect/Ibis/IbisOps.h"
+#include "circt/Dialect/Ibis/IbisPasses.h"
+#include "circt/Dialect/Ibis/IbisTypes.h"
+#include "circt/Dialect/SSP/SSPOps.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include "circt/Transforms/Passes.h"
+
+#include <iterator>
+
+using namespace mlir;
+using namespace circt;
+using namespace ibis;
+
+namespace {
+
+struct AddOperatorLibraryPass
+    : public IbisAddOperatorLibraryBase<AddOperatorLibraryPass> {
+  void runOnOperation() override;
+};
+
+// TODO @mortbopet: once we have c++20, change  this to a templated lambda to
+// avoid passing the builder around.
+template <typename TOp>
+void addOperator(ImplicitLocOpBuilder &b, int latency) {
+  b.create<ssp::OperatorTypeOp>(
+      b.getStringAttr(TOp::getOperationName()),
+      b.getArrayAttr({b.getAttr<ssp::LatencyAttr>(latency)}));
+}
+
+} // anonymous namespace
+
+void AddOperatorLibraryPass::runOnOperation() {
+  auto b = ImplicitLocOpBuilder::atBlockBegin(getOperation().getLoc(),
+                                              getOperation().getBody());
+  auto opLib = b.create<ssp::OperatorLibraryOp>();
+  opLib.setSymNameAttr(b.getStringAttr(kIbisOperatorLibName));
+  b.setInsertionPointToStart(opLib.getBodyBlock());
+
+  // Provide definitions for some comb ops - just latency properties for now.
+
+  // Arithmetic operators
+  addOperator<comb::AddOp>(b, 1);
```

Comment:
The problem with static cost model is that 3-input adds have higher cost then 2 (or 1). Cost functions need to account for that. (I don't know if that's supported by the API, though IIRC it is. Or I at least commented that it should be in a PR years ago.)

Reply:
I suppose one could write out operator types for all the different number of inputs detected. But that gets ugly.

Reply:
fully agree - there are many things in this model (ssp library op/the existing pipeline scheduling pass) that obviously doesn't scale nor will produce good QoR for "real" circuits.